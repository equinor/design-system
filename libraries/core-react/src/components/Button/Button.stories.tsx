import { Button, Icon, ButtonProps, EdsProvider } from '../..'
import styled from 'styled-components'
import { Meta, Story } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { menu } from '@equinor/eds-icons'

const Wrapper = styled.div`
  margin: 32px;
  display: grid;
  grid-gap: 32px;
  grid-template-columns: repeat(4, fit-content(100%));
`

const FullWidthWrapper = styled.div`
  margin: 32px;
  display: grid;
  grid-gap: 16px;
`

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    as: {
      control: {
        type: 'select',
        options: ['span', 'a', 'button'],
        defaultValue: 'button',
      },
    },
  },
  parameters: {
    docs: {
      description: {
        component: `Buttons allow users to take action with a single click or tap.
        `,
      },
    },
    info: {},
  },
} as Meta

export const Default: Story<ButtonProps> = (args) => (
  <Button {...args}>You can control me</Button>
)

export const All: Story<ButtonProps> = () => (
  <Wrapper>
    <Button>Primary</Button>
    <Button color="secondary">Secondary</Button>
    <Button color="danger">Danger</Button>
    <Button disabled>Disabled</Button>
    <Button variant="outlined">Primary</Button>
    <Button variant="outlined" color="secondary">
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      Danger
    </Button>
    <Button variant="outlined" disabled>
      Disabled
    </Button>
    <Button variant="ghost">Primary</Button>
    <Button variant="ghost" color="secondary">
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      Danger
    </Button>
    <Button variant="ghost" disabled>
      Disabled
    </Button>
    <Button variant="ghost_icon">
      <Icon name="save" title="save action"></Icon>
    </Button>
    <Button variant="ghost_icon" color="secondary">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" color="danger">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" disabled>
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Contained: Story<ButtonProps> = () => (
  <Wrapper>
    <Button>Primary</Button>
    <Button color="secondary">Secondary</Button>
    <Button color="danger">Danger</Button>
    <Button disabled>Disabled</Button>
    <Button>
      <Icon name="save" title="save"></Icon>Primary
    </Button>
    <Button color="secondary">
      <Icon name="save" title="save"></Icon>Secondary
    </Button>
    <Button color="danger">
      <Icon name="save" title="save"></Icon>Danger
    </Button>
    <Button disabled>
      <Icon name="save" title="save"></Icon>Disabled
    </Button>
    <Button>
      Primary <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="secondary">
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="danger">
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button disabled>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Outlined: Story<ButtonProps> = () => (
  <Wrapper>
    <Button variant="outlined">Primary</Button>
    <Button variant="outlined" color="secondary">
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      Danger
    </Button>
    <Button variant="outlined" disabled>
      Disabled
    </Button>
    <Button variant="outlined">
      <Icon name="save" title="save"></Icon>
      Primary
    </Button>
    <Button variant="outlined" color="secondary">
      <Icon name="save" title="save"></Icon>
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      <Icon name="save" title="save"></Icon>
      Danger
    </Button>
    <Button variant="outlined" disabled>
      <Icon name="save" title="save"></Icon>
      Disabled
    </Button>
    <Button variant="outlined">
      Primary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="outlined" color="secondary">
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="outlined" color="danger">
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="outlined" disabled>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Ghost: Story<ButtonProps> = () => (
  <Wrapper>
    <Button variant="ghost">Primary</Button>
    <Button variant="ghost" color="secondary">
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      Danger
    </Button>
    <Button variant="ghost" disabled>
      Disabled
    </Button>
    <Button variant="ghost">
      <Icon name="save" title="save"></Icon>
      Primary
    </Button>
    <Button variant="ghost" color="secondary">
      <Icon name="save" title="save"></Icon>
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      <Icon name="save" title="save"></Icon>
      Danger
    </Button>
    <Button variant="ghost" disabled>
      <Icon name="save" title="save"></Icon>
      Disabled
    </Button>
    <Button variant="ghost">
      Primary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost" color="secondary">
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost" color="danger">
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost" disabled>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const GhostIcon: Story<ButtonProps> = () => (
  <Wrapper>
    <Button variant="ghost_icon">
      <Icon name="save" title="save action"></Icon>
    </Button>
    <Button variant="ghost_icon" color="secondary">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" color="danger">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" disabled>
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Form: Story<ButtonProps> = () => {
  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault() // to prevent navigation from storybook
    action('onSubmit')(e)
  }

  return (
    <Wrapper>
      <form onSubmit={handleSubmit}>
        <Button type="submit">Submit form</Button>
      </form>
    </Wrapper>
  )
}

export const FileUpload: Story<ButtonProps> = () => (
  <Wrapper>
    <input type="file" id="file-upload" style={{ display: 'none' }} multiple />
    <label htmlFor="file-upload">
      <Button as="span">Upload</Button>
    </label>
  </Wrapper>
)

FileUpload.parameters = {
  docs: {
    description: {
      story:
        'Please note this demo only works in Storybook Canvas (isolated example only)',
    },
  },
}

export const Link: Story<ButtonProps> = () => (
  <Wrapper>
    <Button href="#">Link</Button>
  </Wrapper>
)

export const DisabledLink: Story<ButtonProps> = () => {
  const isLinkValid = false
  return (
    <Wrapper>
      <Button
        href={isLinkValid ? '#' : undefined}
        disabled={isLinkValid ? false : true}
      >
        Disabled Link
      </Button>
    </Wrapper>
  )
}

DisabledLink.parameters = {
  docs: {
    description: {
      story:
        "Disabling a `Button` with a valid `href` URL does not work, as it's not possible to disable links. A workaround for this is to set `href={undefined}`.",
    },
  },
}

export const FullWidth: Story<ButtonProps> = () => (
  <FullWidthWrapper>
    <Button fullWidth>Primary</Button>
    <Button color="secondary" fullWidth>
      Secondary
    </Button>
    <Button color="danger" fullWidth>
      Danger
    </Button>
    <Button disabled fullWidth>
      Disabled
    </Button>
    <Button fullWidth>
      <Icon name="save" title="save"></Icon>Primary
    </Button>
    <Button color="secondary" fullWidth>
      <Icon name="save" title="save"></Icon>Secondary
    </Button>
    <Button color="danger" fullWidth>
      <Icon name="save" title="save"></Icon>Danger
    </Button>
    <Button disabled fullWidth>
      <Icon name="save" title="save"></Icon>Disabled
    </Button>
    <Button fullWidth>
      Primary <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="secondary" fullWidth>
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="danger" fullWidth>
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button disabled fullWidth>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </FullWidthWrapper>
)

export const Compact: Story<ButtonProps> = () => (
  <EdsProvider density="compact">
    <Wrapper>
      <Button>Contained</Button>
      <Button variant="outlined">Outlined</Button>
      <Button variant="ghost">Ghost</Button>
      <Button variant="ghost_icon">
        <Icon data={menu} title="Ghost icon menu"></Icon>
      </Button>
    </Wrapper>
  </EdsProvider>
)

Compact.parameters = {
  docs: {
    description: {
      story: 'Compact `Button` using `EdsProvider`',
    },
  },
}
