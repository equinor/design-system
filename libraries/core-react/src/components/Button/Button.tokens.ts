import { tokens } from '@equinor/eds-tokens'
import type {
  Border,
  Focus,
  Hover,
  Spacing,
  Typography,
} from '@equinor/eds-tokens'

const {
  components: {
    button: { primary, secondary, danger, disabled },
  },
} = tokens

const colors = {
  primary: {
    ...primary,
    ghost_icon: {
      ...primary.ghost_icon,
      typography: primary.ghost.typography,
    },
  },
  secondary: {
    ...secondary,
    ghost_icon: {
      ...secondary.ghost_icon,
      typography: secondary.ghost.typography,
    },
  },
  danger: {
    ...danger,
    ghost_icon: {
      ...danger.ghost_icon,
      typography: danger.ghost.typography,
    },
  },
  disabled: {
    ...disabled,
    ghost_icon: {
      ...disabled.ghost_icon,
      typography: disabled.ghost.typography,
    },
  },
}

export type Button = {
  height: string
  width?: string
  background: string
  color: string
  border: Border
  typography: Typography
  spacing?: Partial<Spacing>
  focus?: Focus
  hover?: Hover
  // TODO Remove these once figma-broker is updated with proper types
  pressedColor?: string
  clickboundOffset?: number | string
  clickbound?: string
}

type Buttons = {
  [P in keyof typeof colors]: {
    [P2 in keyof typeof colors[P]]: Button
  }
}

export type ButtonGroups =
  | Buttons['primary']
  | Buttons['secondary']
  | Buttons['danger']
  | Buttons['disabled']

export const button: {
  colors: Buttons
  icon_size: {
    width: string
    height: string
  }
} = {
  colors,
  icon_size: {
    width: '24px',
    height: '24px',
  },
}
