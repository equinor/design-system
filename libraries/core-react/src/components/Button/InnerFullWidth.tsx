import { forwardRef, Children as ReactChildren, HTMLAttributes } from 'react'
import styled from 'styled-components'

const FullWidthCenterContent = styled.span`
  text-align: center;
  flex: 1;
`

const FullWidthInner = styled.span`
  height: 100%;
  display: flex;
  align-items: center;

  > img:first-child,
  > svg:first-child {
    margin-right: 8px;
  }

  > img:last-child,
  > svg:last-child {
    margin-left: 8px;
  }

  > img:only-child,
  > svg:only-child {
    margin-left: auto;
    margin-right: auto;
  }

  > span:first-child {
    margin-left: 32px;
  }

  > span:last-child {
    margin-right: 32px;
  }

  > span:only-child {
    margin-right: 0;
    margin-left: 0;
  }
`

export const InnerFullWidth = forwardRef<
  HTMLSpanElement,
  HTMLAttributes<HTMLSpanElement>
>(function InnerFullWidth({ children }) {
  // We need everything in elements for proper flexing ðŸ’ª
  const updatedChildren = ReactChildren.map(children, (child) =>
    typeof child !== 'object' ? (
      <FullWidthCenterContent>{child}</FullWidthCenterContent>
    ) : (
      child
    ),
  )

  return <FullWidthInner>{updatedChildren}</FullWidthInner>
})
