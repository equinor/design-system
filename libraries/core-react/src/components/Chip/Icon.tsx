import styled, { css } from 'styled-components'
import { close } from '@equinor/eds-icons'
import { Icon as Icon_ } from '../Icon'
import * as tokens from './Chip.tokens'
import { bordersTemplate } from '@utils'

Icon_.add({ close })

const { primary, error } = tokens

type IconProps = {
  variant: 'active' | 'error' | 'default'
  disabled: boolean
}

export const Icon = styled(Icon_)<IconProps>`
  cursor: pointer;
  padding: 1px;
  ${bordersTemplate(primary.entities.icon.border)}

  z-index: 11;

  &:hover {
    ${({ variant }) => {
      switch (variant) {
        case 'error':
          return css`
            background: ${error.entities.icon.background};
            svg {
              fill: ${error.entities.icon.typography.color};
            }
          `
        default:
          return css`
            background: ${primary.entities.icon.states.hover.background};
            svg {
              fill: ${primary.states.hover.typography.color};
            }
          `
      }
    }};
  }

  ${({ disabled }) =>
    disabled &&
    css`
      cursor: not-allowed;
      &:hover {
        background: transparent;
      }
    `}
`
