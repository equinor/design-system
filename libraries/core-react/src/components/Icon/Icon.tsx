import * as React from 'react'
import { forwardRef, Ref, SVGProps } from 'react'
import styled from 'styled-components'
import { get } from './library'
import type { IconData } from '@equinor/eds-icons'
import type { IconBasket, Name } from './Icon.types'

type StyledProps = {
  height: number
  width: number
  fill: string
  size?: number
  rotation?: number
}

type SvgProps = {
  name: string
  viewBox: string
  className: string
  rotation?: number
  title?: string
  role?: string
  'aria-hidden'?: boolean | 'true' | 'false'
  'aria-labelledby'?: string
}

const customIcon = (icon: IconData): IconBasket => ({
  icon,
  count: Math.floor(Math.random() * 1000),
})

const transform = ({ rotation }: SvgProps): string =>
  rotation ? `transform: rotate(${rotation}deg)` : ''

const StyledSvg = styled.svg.attrs<StyledProps>(({ height, width, fill }) => ({
  name: null,
  xmlns: 'http://www.w3.org/2000/svg',
  height: `${height}px`,
  width: `${width}px`,
  fill,
}))`
  ${transform}
`

const StyledPath = styled.path.attrs<StyledProps>(({ height, size }) => ({
  size: null,
  fillRule: 'evenodd',
  clipRule: 'evenodd',
  transform: size / height !== 1 ? `scale(${size / height})` : null,
}))``

export type IconProps = {
  /** Title for icon when used semantically */
  title?: string
  /** Color */
  color?: string
  /** Size */
  size?: 16 | 24 | 32 | 40 | 48
  /** Rotation */
  rotation?: 0 | 90 | 180 | 270
  /** Name */
  name?: Name
  /** Manually specify which icon data to use */
  data?: IconData
  /** @ignore */
  ref?: Ref<SVGSVGElement>
} & SVGProps<SVGSVGElement>

export const Icon = forwardRef<SVGSVGElement, IconProps>(function Icon(
  {
    size = 24,
    color = 'currentColor',
    name,
    className,
    rotation,
    title,
    data,
    ...rest
  },
  ref,
) {
  const { icon, count }: IconBasket = data ? customIcon(data) : get(name)

  if (typeof icon === 'undefined') {
    throw Error(
      `Icon "${name}" not found. Have you added it using Icon.add() or using data props?`,
    )
  }

  let svgProps: SvgProps & StyledProps = {
    height: size,
    width: size,
    fill: color,
    viewBox: `0 0 ${size} ${size}`,
    className,
    rotation,
    name,
    'aria-hidden': true,
    title: null,
  }

  const pathProps = {
    d: icon.svgPathData,
    height: icon.height ? icon.height : size,
    size,
  }

  // Accessibility
  let titleId = ''

  if (title) {
    titleId = `${icon.prefix}-${icon.name}-${count}`
    svgProps = {
      ...svgProps,
      title,
      role: 'img',
      'aria-hidden': undefined,
      'aria-labelledby': titleId,
    }
  }

  return (
    <StyledSvg {...svgProps} {...rest} ref={ref}>
      {title && <title id={titleId}>{title}</title>}
      <StyledPath data-testid="eds-icon-path" {...pathProps} />
    </StyledSvg>
  )
})

// Icon.displayName = 'EdsIcon'
