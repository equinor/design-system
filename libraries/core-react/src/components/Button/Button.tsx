import { forwardRef, ElementType, ButtonHTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { token as buttonToken } from './tokens'
import { ButtonTokenSet, ButtonToken } from './Button.types'
import {
  typographyTemplate,
  bordersTemplate,
  outlineTemplate,
  spacingsTemplate,
} from '../../utils'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '../EdsProvider'

type Colors = 'primary' | 'secondary' | 'danger'
type Variants = 'contained' | 'outlined' | 'ghost' | 'ghost_icon'

const getVariant = (
  tokenSet: ButtonTokenSet,
  variant: Variants,
): ButtonToken => {
  switch (variant) {
    case 'ghost':
      return tokenSet.ghost
    case 'ghost_icon':
      return tokenSet.ghost_icon
    case 'outlined':
      return tokenSet.outlined
    case 'contained':
    default:
      return tokenSet.contained
  }
}

const getToken = (variant: Variants, color: Colors): ButtonToken => {
  switch (color) {
    case 'danger':
      return getVariant(buttonToken.danger, variant)
    case 'secondary':
      return getVariant(buttonToken.secondary, variant)
    case 'primary':
    default:
      return getVariant(buttonToken.primary, variant)
  }
}

const ButtonInner = styled.span`
  display: grid;
  grid-gap: 8px;
  grid-auto-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;
`

const Base = ({ token, density }: { token: ButtonToken; density: string }) => {
  const { spacings, states } = token
  const { focus, hover, disabled } = states

  let height = token.height
  let width = token.width
  let clickbound = token.clickbound

  if (density === 'compact') {
    height = token.modes.compact.height
    width = token.modes.compact.width
    clickbound = token.modes.compact.clickbound
  }

  return css`
    background: ${token.background};
    height: ${height};
    width: ${width};
    svg {
      justify-self: center;
    }

    ${spacingsTemplate(spacings)}
    ${bordersTemplate(token.border)}
    ${typographyTemplate(token.typography)}

    &::after {
      position: absolute;
      top: -${clickbound?.offset?.top};
      left: -${clickbound?.offset?.left};
      width: ${clickbound?.width};
      height: ${clickbound?.height};
      content: '';
    }

    &:hover {
      background: ${hover.background};
      color: ${hover.typography?.color};
      ${bordersTemplate(hover?.border)}
    }

    &:focus {
      outline: none;
    }

    &[data-focus-visible-added]:focus {
      ${outlineTemplate(focus.outline)}
    }
    /* Get rid of ff focus border for buttons */
    &::-moz-focus-inner {
      border: 0;
    }

    &:disabled {
      cursor: not-allowed;
      background: ${disabled.background};
      ${bordersTemplate(disabled.border)}
      ${typographyTemplate(disabled.typography)}

      &:hover {
        background: ${disabled.background};
      }
    }
  `
}

const ButtonBase = styled.button`
  margin: 0;
  padding: 0;
  ${Base}
  text-decoration: none;
  position: relative;
  cursor: pointer;
  display: inline-block;

  &::before {
    position: absolute;
    top: 0;
    left: 0;
    width: auto;
    min-height: auto;
    content: '';
  }
`
export type ButtonProps = {
  /**  Specifies color */
  color?: 'primary' | 'secondary' | 'danger'
  /** Specifies which variant to use */
  variant?: 'contained' | 'outlined' | 'ghost' | 'ghost_icon'
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Change html element. */
  as?: ElementType
  /** Type of button
   * @default 'button'
   */
  type?: string
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      ...other
    },
    ref,
  ) {
    const token = getToken(variant, color)
    const { density } = useEds()
    const as: ElementType =
      href && !disabled ? 'a' : other.as ? other.as : 'button'
    const type = href || other.as ? undefined : 'button'

    tabIndex = disabled ? -1 : tabIndex

    const buttonProps = {
      ref,
      as,
      href,
      type,
      token,
      disabled,
      tabIndex,
      density,
      ...other,
    }

    return (
      <ButtonBase {...buttonProps}>
        {fullWidth ? (
          <InnerFullWidth>{children}</InnerFullWidth>
        ) : (
          <ButtonInner>{children}</ButtonInner>
        )}
      </ButtonBase>
    )
  },
)
