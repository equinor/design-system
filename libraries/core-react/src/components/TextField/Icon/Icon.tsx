import * as React from 'react'
import { HTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { useTextField } from '../context'
import { input as tokens } from './Icon.token'
import type { Spacing } from '@equinor/eds-tokens'
import type { Variants, ColorStateProps } from '../types'

type StyledIconProps = {
  colors: ColorStateProps
  isDisabled?: boolean
  isFocused?: boolean
  spacings: Spacing
}

const Variation = ({ colors, isDisabled, isFocused }: StyledIconProps) => {
  const { focusColor, color, disabledColor } = colors

  if (isDisabled) {
    return css`
      &,
      svg {
        fill: ${disabledColor};
      }
    `
  }
  if (isFocused) {
    return css`
      &,
      svg {
        fill: ${focusColor};
      }
    `
  }
  return css`
    &,
    svg {
      fill: ${color};
    }
  `
}

const StyledIcon = styled.div<StyledIconProps>`
  &,
  svg {
    width: 16px;
    height: 16px;
  }
  ${Variation}
`

const StyledIputIcon = styled(StyledIcon)`
  position: absolute;
  right: ${({ spacings }) => spacings.right};
  top: ${({ spacings }) => spacings.top};
  bottom: ${({ spacings }) => spacings.bottom};
`

type TextfieldIconProps = {
  /** isDisabled */
  isDisabled?: boolean
  /** Variant */
  variant?: Variants
  /** Is the icon inside a text field */
  isInputIcon?: boolean
  /** Spacing object, comfortable is default */
  spacings?: Spacing
  /** Colors */
  colors?: ColorStateProps
} & HTMLAttributes<HTMLDivElement>

const InputIcon = React.forwardRef<HTMLDivElement, TextfieldIconProps>(
  function InputIcon(
    {
      variant = 'default',
      isDisabled = false,
      isInputIcon = true,
      spacings = tokens.spacings.comfortable,
      colors = {
        color: tokens[variant].color,
        disabledColor: tokens[variant].disabledColor,
        focusColor: tokens[variant].focusColor,
      },
      children,
      ...other
    },
    ref,
  ) {
    const { isFocused } = useTextField()

    const iconProps = {
      spacings,
      isDisabled,
      colors,
      isFocused,
    }

    return (
      <>
        {isInputIcon ? (
          <StyledIputIcon ref={ref} {...iconProps} {...other}>
            {children}
          </StyledIputIcon>
        ) : (
          <StyledIcon ref={ref} {...iconProps} {...other}>
            {children}
          </StyledIcon>
        )}
      </>
    )
  },
)

export { InputIcon as Icon }
