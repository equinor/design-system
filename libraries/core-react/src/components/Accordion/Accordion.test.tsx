/* eslint-disable no-undef */
import { MouseEventHandler } from 'react'
import { render, cleanup, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
// eslint-disable-next-line camelcase
import { attach_file, notifications } from '@equinor/eds-icons'
import { Accordion } from '.'
import { Button } from '../Button'
import { Icon } from '../Icon'
import type { AccordionProps } from './Accordion.types'
import React from 'react'

Icon.add({ attach_file, notifications })

afterEach(cleanup)

const SimpleAccordion = ({
  headerLevel = 'h2',
  chevronPosition = 'left',
}: AccordionProps) => (
  <Accordion headerLevel={headerLevel} chevronPosition={chevronPosition}>
    <Accordion.Item isExpanded>
      <Accordion.Header>Summary 1</Accordion.Header>
      <Accordion.Panel>Details 1</Accordion.Panel>
    </Accordion.Item>
    <Accordion.Item>
      <Accordion.Header>Summary 2</Accordion.Header>
      <Accordion.Panel>Details 2</Accordion.Panel>
    </Accordion.Item>
  </Accordion>
)

const AccordionWithIcons = () => (
  <Accordion>
    <Accordion.Item>
      <Accordion.Header>
        <Accordion.HeaderTitle>Summary</Accordion.HeaderTitle>
        <Icon name="attach_file" title="Attach file" size={16} />
        <Icon name="notifications" title="Notifications" size={16} />
      </Accordion.Header>
      <Accordion.Panel>Details</Accordion.Panel>
    </Accordion.Item>
  </Accordion>
)

const AccordionWithButtons = () => (
  <Accordion>
    <Accordion.Item>
      <Accordion.Header>
        <Accordion.HeaderTitle>Summary</Accordion.HeaderTitle>
        <Button
          variant="ghost_icon"
          onClick={(event: React.MouseEvent<HTMLButtonElement, MouseEvent>) =>
            event.stopPropagation()
          }
        >
          <Icon name="attach_file" title="Attach file" />
        </Button>
        <Button
          variant="ghost_icon"
          onClick={(event: React.MouseEvent<HTMLButtonElement, MouseEvent>) =>
            event.stopPropagation()
          }
        >
          <Icon name="notifications" title="Notifications" />
        </Button>
      </Accordion.Header>
      <Accordion.Panel>Details</Accordion.Panel>
    </Accordion.Item>
  </Accordion>
)

describe('Accordion', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<SimpleAccordion />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Expands items based on prop', () => {
    render(<SimpleAccordion />)
    const header1 = screen.queryByText('Summary 1').parentNode
    const header2 = screen.queryByText('Summary 2').parentNode
    expect(header1).toHaveAttribute('aria-expanded', 'true')
    expect(header2).toHaveAttribute('aria-expanded', 'false')
  })
  it('Expands items on click', () => {
    render(<SimpleAccordion />)
    const header = screen.queryByText('Summary 2').parentNode
    fireEvent.click(header)
    expect(header).toHaveAttribute('aria-expanded', 'true')
  })
  it('triggers onToggle callback', () => {
    const mockOnToggle = jest.fn()
    render(
      <Accordion.Item isExpanded>
        <Accordion.Header onToggle={mockOnToggle}>Summary 1</Accordion.Header>
      </Accordion.Item>,
    )
    const header = screen.queryByText('Summary 1').parentNode
    fireEvent.click(header)
    expect(mockOnToggle).toHaveBeenCalled()
  })
  it('Set header level', () => {
    render(<SimpleAccordion headerLevel="h3" />)
    expect(document.querySelectorAll('h3')).toHaveLength(2)
  })
  it('Has chevron on left side as default', () => {
    render(<SimpleAccordion />)
    const header = screen.queryByText('Summary 1').parentNode
    const chevron = header.querySelector('svg')
    expect(header.firstChild).toBe(chevron)
  })
  it('Set chevron position to the right', () => {
    render(<SimpleAccordion chevronPosition="right" />)
    const header = screen.queryByText('Summary 1').parentNode
    const chevron = header.querySelector('svg')
    expect(header.lastChild).toBe(chevron)
  })
  it('Add custom icons', () => {
    render(<AccordionWithIcons />)
    const header = screen.queryByText('Summary').parentNode
    expect(header.querySelectorAll('svg')).toHaveLength(3)
  })
  it('Add custom buttons', () => {
    render(<AccordionWithButtons />)
    const header = screen.queryByText('Summary').parentNode
    expect(header.querySelectorAll('button')).toHaveLength(2)
  })
  it('Does not expand when clicking custom buttons', () => {
    render(<AccordionWithButtons />)
    const header = screen.queryByText('Summary').parentNode
    const button = screen.getAllByTitle('Notifications')[0].parentNode
    fireEvent.click(button)
    expect(header).toHaveAttribute('aria-expanded', 'false')
  })
})
