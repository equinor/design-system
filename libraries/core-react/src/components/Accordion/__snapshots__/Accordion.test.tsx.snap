// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Matches snapshot 1`] = `
<DocumentFragment>
  .c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  text-align: left;
  font-family: Equinor;
  color: var(--eds_interactive_primary__resting,rgba(0,112,121,1));
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  text-align: left;
  font-family: Equinor;
  color: inherit;
}

.c0 {
  margin: 0;
  color: rgba(61,61,61,1);
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 500;
  line-height: 1.500em;
  text-align: left;
  border: 1px solid var(--eds_ui_background__medium,rgba(220,220,220,1));
  padding-left: 16px;
  padding-right: 16px;
  background: var(--eds_ui_background__default,rgba(255,255,255,1));
  height: 48px;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  color: rgba(61,61,61,1);
  cursor: pointer;
}

.c0[data-focus-visible-added]:focus {
  outline: 1px dashed var(--eds_interactive__focus,rgba(0,112,121,1));
  outline-offset: 2px;
}

.c0:hover {
  background: var(--eds_ui_background__light,rgba(247,247,247,1));
}

.c4 {
  margin: 0;
  color: rgba(61,61,61,1);
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 500;
  line-height: 1.500em;
  text-align: left;
  border: 1px solid var(--eds_ui_background__medium,rgba(220,220,220,1));
  padding-left: 16px;
  padding-right: 16px;
  border-top: none;
  background: var(--eds_ui_background__default,rgba(255,255,255,1));
  height: 48px;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  color: rgba(61,61,61,1);
  cursor: pointer;
}

.c4[data-focus-visible-added]:focus {
  outline: 1px dashed var(--eds_interactive__focus,rgba(0,112,121,1));
  outline-offset: 2px;
}

.c4:hover {
  background: var(--eds_ui_background__light,rgba(247,247,247,1));
}

.c1 {
  margin-right: 32px;
}

.c3 {
  border: 1px solid var(--eds_ui_background__medium,rgba(220,220,220,1));
  padding-left: 16px;
  padding-top: 16px;
  padding-right: 16px;
  padding-bottom: 16px;
  background: var(--eds_ui_background__default,rgba(255,255,255,1));
  min-height: 96px;
  box-sizing: border-box;
}

<div>
    <div>
      <h2
        aria-controls="accordion-1-panel-1"
        aria-expanded="true"
        class="c0"
        role="button"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="c1"
          fill="var(--eds_interactive_primary__resting, rgba(0, 112, 121, 1))"
          height="16px"
          viewBox="0 0 16 16"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            clip-rule="evenodd"
            d="M7.41 15.705l4.59-4.58 4.59 4.58 1.41-1.41-6-6-6 6 1.41 1.41z"
            data-testid="eds-icon-path"
            fill-rule="evenodd"
            height="24"
            transform="scale(0.6666666666666666)"
          />
        </svg>
        <span
          class="c2"
        >
          Summary 1
        </span>
      </h2>
      <div
        aria-labelledby="accordion-1-header-1"
        class="c3"
        id="accordion-1-panel-1"
        role="region"
      >
        Details 1
      </div>
    </div>
    <div>
      <h2
        aria-controls="accordion-1-panel-2"
        aria-disabled="false"
        aria-expanded="false"
        class="c4"
        role="button"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="c1"
          fill="var(--eds_interactive_primary__resting, rgba(0, 112, 121, 1))"
          height="16px"
          viewBox="0 0 16 16"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            clip-rule="evenodd"
            d="M7.41 8.295l4.59 4.58 4.59-4.58L18 9.705l-6 6-6-6 1.41-1.41z"
            data-testid="eds-icon-path"
            fill-rule="evenodd"
            height="24"
            transform="scale(0.6666666666666666)"
          />
        </svg>
        <span
          class="c5"
        >
          Summary 2
        </span>
      </h2>
      <div
        aria-labelledby="accordion-1-header-2"
        class="c3"
        hidden=""
        id="accordion-1-panel-2"
        role="region"
      >
        Details 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;
